e2e-tests/upgrade-haproxy/run

```diff
@@ -0,0 +1,268 @@
+#!/bin/bash
+
+set -o errexit
+set -o xtrace
+
+test_dir=$(realpath $(dirname $0))
+. ${test_dir}/../functions
+
+CLUSTER="upgrade-haproxy"
+CLUSTER_SIZE=3
+TARGET_API="${API}"
+TARGET_OPERATOR_VER=$(echo -n "${API//pxc.percona.com\/v}" | $sed 's/-/./g')
+TARGET_IMAGE="${IMAGE}"
+TARGET_IMAGE_PXC="${IMAGE_PXC}"
+TARGET_IMAGE_PMM="${IMAGE_PMM}"
+TARGET_IMAGE_PROXY="${IMAGE_PROXY}"
+TARGET_IMAGE_HAPROXY="${IMAGE_HAPROXY}"
+TARGET_IMAGE_BACKUP="${IMAGE_BACKUP}"
+if [[ "${TARGET_IMAGE_PXC}" == *"percona-xtradb-cluster-operator"* ]]; then
+    PXC_VER=$(echo -n "${TARGET_IMAGE_PXC}" | $sed -r 's/.*([0-9].[0-9])$/\1/')
+else
+    PXC_VER=$(echo -n "${TARGET_IMAGE_PXC}" | $sed -r 's/.*:([0-9]+\.[0-9]+).*$/\1/')
+fi
+
+INIT_OPERATOR_VER=$(curl -s https://check.percona.com/versions/v1/pxc-operator | jq -r '.versions[].operator' | sort -V | tail -n1)
+# if testing on release branch and version service is already updated with new operator images
+# use the older version of operator as initial point for test
+if [[ "${INIT_OPERATOR_VER}" == "${TARGET_OPERATOR_VER}" ]]; then
+    INIT_OPERATOR_VER=$(curl -s https://check.percona.com/versions/v1/pxc-operator | jq -r '.versions[].operator' | sort -V | tail -n2 | head -n1)
+fi
+GIT_TAG="v${INIT_OPERATOR_VER}"
+INIT_OPERATOR_IMAGES=$(curl -s "https://check.percona.com/versions/v1/pxc-operator/${INIT_OPERATOR_VER}/latest?databaseVersion=${PXC_VER}")
+OPERATOR_NAME="percona-xtradb-cluster-operator"
+API="pxc.percona.com/v${INIT_OPERATOR_VER//./-}"
+IMAGE=$(echo "${INIT_OPERATOR_IMAGES}" | jq -r '.versions[].matrix.operator[].imagePath')
+# we use the starting image from the same repo so we don't need to use initImage option
+if [[ "$(echo ${TARGET_IMAGE}|cut -d'/' -f1)" == "perconalab" ]]; then
+    IMAGE="${IMAGE/percona/perconalab}"
+fi
+IMAGE_PXC=$(echo "${INIT_OPERATOR_IMAGES}" | jq -r '.versions[].matrix.pxc[].imagePath')
+IMAGE_PMM=$(echo "${INIT_OPERATOR_IMAGES}" | jq -r '.versions[].matrix.pmm[].imagePath')
+IMAGE_PROXY=$(echo "${INIT_OPERATOR_IMAGES}" | jq -r '.versions[].matrix.proxysql[].imagePath')
+IMAGE_HAPROXY=$(echo "${INIT_OPERATOR_IMAGES}" | jq -r '.versions[].matrix.haproxy[].imagePath')
+IMAGE_BACKUP=$(echo "${INIT_OPERATOR_IMAGES}" | jq -r '.versions[].matrix.backup[].imagePath')
+
+if [[ "${TARGET_API}" == "${API}" ]]; then
+    echo "API and TARGET_API variables are the same: ${API}! Something is wrong!"
+    exit 1
+fi
+
+function check_pxc_liveness() {
+    wait_cluster_consistency "${CLUSTER}" "${CLUSTER_SIZE}"
+
+    wait_for_running "${CLUSTER}-pxc" "${CLUSTER_SIZE}"
+
+    for i in $(seq 0 $((CLUSTER_SIZE-1))); do
+        compare_mysql_cmd "select-1" "SELECT * from myApp.myApp;" "-h ${CLUSTER}-pxc-${i}.${CLUSTER}-pxc -uroot -proot_password"
+    done
+}
+
+function compare_generation() {
```

Comment:
we use the same function in two different tests, we need to move it to functions

Reply:
done